File: flowClient.js
// flowClient.js
// Cliente simple para crear pagos en Flow desde tu backend (Node ESM)

import crypto from "crypto";

// ====== ENV ======
// Debes tener definidas en Render (o donde sea tu backend):
//
// FLOW_API_KEY      → la API Key que ves en el panel de Flow
// FLOW_SECRET_KEY   → la Secret Key que ves en el panel de Flow
// FLOW_ENV          → "sandbox" o "production" (opcional, por defecto sandbox)

const FLOW_API_KEY = process.env.FLOW_API_KEY || "";
const FLOW_SECRET_KEY = process.env.FLOW_SECRET_KEY || "";
const FLOW_ENV = process.env.FLOW_ENV || "sandbox";

const FLOW_BASE_URL =
  FLOW_ENV === "production"
    ? "https://www.flow.cl/api"
    : "https://sandbox.flow.cl/api";

// Pequeña validación de entorno
function assertEnv() {
  if (!FLOW_API_KEY || !FLOW_SECRET_KEY) {
    throw new Error(
      "Faltan variables de entorno de Flow (FLOW_API_KEY / FLOW_SECRET_KEY)."
    );
  }
}

// Construye la firma HMAC-SHA256 según la documentación de Flow.
// IMPORTANTE: la firma se hace sobre los parámetros ordenados por nombre,
// sin incluir el campo `s`.
function makeSignature(params) {
  const orderedKeys = Object.keys(params).sort();
  const queryString = orderedKeys
    .map((k) => `${k}=${params[k]}`)
    .join("&");

  return crypto
    .createHmac("sha256", FLOW_SECRET_KEY)
    .update(queryString)
    .digest("hex");
}

// Helper para construir URLSearchParams (Flow espera application/x-www-form-urlencoded)
function toFormBody(params) {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null) continue;
    sp.append(k, String(v));
  }
  return sp.toString();
}

/**
 * Crear pago en Flow.
 *
 * @param {Object} opts
 * @param {string} opts.idPago        - identificador interno (lo usamos como commerceOrder)
 * @param {number} opts.amount        - monto en CLP (ej: 1000)
 * @param {string} opts.subject       - texto corto, ejemplo: "Orden Asistencia ICA"
 * @param {string} opts.email         - correo del paciente (o tuyo si no tienes)
 * @param {string} opts.modulo        - "trauma" | "preop" | "generales" | "ia"
 * @param {string} opts.urlConfirmation - URL de confirmación en tu backend
 * @param {string} opts.urlReturn       - URL de retorno al frontend (?pago=...)
 * @param {Object} [opts.optionalData]  - datos adicionales para guardar en Flow (se serializan a JSON)
 *
 * @returns {Promise<{url:string, token:string|null}>}
 */
export async function crearPagoFlowBackend({
  idPago,
  amount,
  subject,
  email,
  modulo = "trauma",
  urlConfirmation,
  urlReturn,
  optionalData = {},
}) {
  assertEnv();

  if (!idPago) throw new Error("crearPagoFlowBackend requiere idPago");
  if (!amount || Number(amount) <= 0)
    throw new Error("crearPagoFlowBackend requiere amount > 0");
  if (!subject) subject = `Orden ${modulo.toUpperCase()} Asistencia ICA`;
  if (!email) email = "sin-correo@icarticular.cl";
  if (!urlConfirmation || !urlReturn) {
    throw new Error(
      "crearPagoFlowBackend requiere urlConfirmation y urlReturn configuradas."
    );
  }

  const baseParams = {
    apiKey: FLOW_API_KEY,
    commerceOrder: idPago,
    subject,
    currency: "CLP",
    amount: Number(amount),
    email,
    urlConfirmation,
    urlReturn,
    // Puedes pasar datos adicionales en `optional` (string JSON)
    optional:
      optionalData && Object.keys(optionalData).length
        ? JSON.stringify({ modulo, ...optionalData })
        : JSON.stringify({ modulo }),
  };

  // Firmamos
  const s = makeSignature(baseParams);
  const payload = { ...baseParams, s };

  const url = `${FLOW_BASE_URL}/payment/create`;

  const res = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Accept: "application/json",
    },
    body: toFormBody(payload),
  });

  const raw = await res.text();
  let data = null;
  try {
    data = JSON.parse(raw);
  } catch {
    // Flow a veces devuelve texto plano en errores
  }

  if (!res.ok || !data || data.status !== 1 || !data.url) {
    const msg =
      data?.message ||
      data?.error ||
      `Error HTTP ${res.status} al crear pago en Flow`;
    throw new Error(`${msg} — Respuesta: ${raw}`);
  }

  return {
    url: data.url,
    token: data.token ?? null,
    flowOrder: data.flowOrder ?? null,
  };
}

File: routers/flow_client.py
# routers/flow_client.py
# Cliente para crear pagos en Flow desde el backend
# Firma HMAC-SHA256 según documentación de Flow
# Stateless — sin memoria server-side

import hashlib
import hmac
import json
import logging
from urllib.parse import urlencode

import httpx

logger = logging.getLogger("flow_client")

FLOW_BASE_SANDBOX    = "https://sandbox.flow.cl/api"
FLOW_BASE_PRODUCTION = "https://www.flow.cl/api"


# ============================================================
# HELPERS
# ============================================================
def _assert_env(config: dict) -> None:
    if not config.get("flow_api_key") or not config.get("flow_secret_key"):
        raise ValueError(
            "Faltan variables de entorno de Flow (FLOW_API_KEY / FLOW_SECRET_KEY)."
        )


def _make_signature(params: dict, secret_key: str) -> str:
    """
    Firma HMAC-SHA256 sobre parámetros ordenados por nombre.
    Sin incluir el campo 's' — igual que el JS original.
    """
    ordered_keys = sorted(params.keys())
    query_string = "&".join(f"{k}={params[k]}" for k in ordered_keys)
    return hmac.new(
        secret_key.encode("utf-8"),
        query_string.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def _to_form_body(params: dict) -> str:
    return urlencode({k: str(v) for k, v in params.items() if v is not None})


def _base_url(config: dict) -> str:
    env = str(config.get("flow_env") or "sandbox").lower()
    return FLOW_BASE_PRODUCTION if env == "production" else FLOW_BASE_SANDBOX


# ============================================================
# CREAR PAGO
# ============================================================
async def crear_pago_flow_backend(
    id_pago:          str,
    amount:           int,
    subject:          str,
    email:            str,
    url_confirmation: str,
    url_return:       str,
    config:           dict,
    modulo:           str = "trauma",
    optional_data:    dict | None = None,
) -> dict:
    """
    Crea un pago en Flow.

    config: { flow_api_key, flow_secret_key, flow_env? }
    Retorna: { url, token, flow_order }
    """
    _assert_env(config)

    if not id_pago:
        raise ValueError("crear_pago_flow_backend requiere id_pago")
    if not amount or int(amount) <= 0:
        raise ValueError("crear_pago_flow_backend requiere amount > 0")
    if not subject:
        subject = f"Orden {modulo.upper()} Asistencia ICA"
    if not email:
        email = "sin-correo@icarticular.cl"
    if not url_confirmation or not url_return:
        raise ValueError(
            "crear_pago_flow_backend requiere url_confirmation y url_return"
        )

    api_key    = config["flow_api_key"]
    secret_key = config["flow_secret_key"]

    optional_dict = {"modulo": modulo}
    if optional_data:
        optional_dict.update(optional_data)

    base_params: dict = {
        "apiKey":          api_key,
        "commerceOrder":   id_pago,
        "subject":         subject,
        "currency":        "CLP",
        "amount":          int(amount),
        "email":           email,
        "urlConfirmation": url_confirmation,
        "urlReturn":       url_return,
        "optional":        json.dumps(optional_dict),
    }

    signature = _make_signature(base_params, secret_key)
    payload   = {**base_params, "s": signature}

    url_endpoint = f"{_base_url(config)}/payment/create"

    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            r = await client.post(
                url_endpoint,
                content=_to_form_body(payload),
                headers={
                    "Content-Type": "application/x-www-form-urlencoded",
                    "Accept":       "application/json",
                },
            )
        raw = r.text
    except httpx.HTTPError as e:
        raise RuntimeError(f"Error de red al crear pago en Flow: {e}") from e

    data = None
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        pass

    # HTTP 2xx + URL → éxito (igual que el JS)
    if r.is_success and data and data.get("url"):
        return {
            "url":        data["url"],
            "token":      data.get("token"),
            "flow_order": data.get("flowOrder"),
        }

    msg = (
        (data or {}).get("message")
        or (data or {}).get("error")
        or f"Error HTTP {r.status_code} al crear pago en Flow"
    )
    raise RuntimeError(f"{msg} — Respuesta: {raw}")
